### ISSUE TITLE
[BUG] Possible fix: Upgrade 1.3.0 -> 1.3.1, last node stuck in pre-drain.

### ISSUE THREAD

**Issue by lidstah** [harvester/harvester#6094]
**Describe the bug**
This afternoon I was upgrading my homelab harvester cluster from 1.3.0 to 1.3.1. 

The cluster has no plugins activated and no special or non-standard configuration options setup.

All VMs were stopped during the upgrade.

The two first nodes upgraded flawlessly, but the last node ended up stalled in Pre-Drain mode. No Pod Disruption Budget were stalling the update (cf [documentation on pre-drain stuck nodes](https://docs.harvesterhci.io/v1.3/upgrade/v1-1-2-to-v1-2-0#5-an-upgrade-is-stuck-in-the-pre-drained-state)). I noticed that the harvester-upgrade VM was failing to launch with the same error message as in [#5755](https://github.com/harvester/harvester/issues/5755).

Looking at running VMs in the harvester-system namespace with `kubectl get vmi -n harvester-system` I noticed many failed harvester-upgrade-ID VMs, and one running VM instance with the same name - which was running since ~45 minutes.

**Possible fix**

I then looked at the pods running in the harvester-system namespace, and noticed an harvester-update-ID-pre-drain-nodename-podID pod with the same ID than the failed (and the running) VM instances. 

Deleting this pod solved the problem, as the cluster spawned a new pod and thus a new VM Instance once the previous pod was properly terminated. The VM Instance launched without problem and the upgrade then proceeded to post-drain and reboot phases without problem.

**To Reproduce**
1. Start the upgrade to 1.3.1 from a 1.3.0 3 nodes cluster.

**Expected behavior**
Upgrade should behave smoothly.

**Support bundle**
Sadly, I forgot to download the support bundle :(. If I encounter the same problem when upgrading from 1.3.1 to 1.3.2 I'll be sure to download it and toroughly document my findings.

**Environment**
 - Harvester 1.3.0 upgrade to 1.3.1
 - Underlying Infrastructure: Baremetal, with 3 NUCs, Intel Core i7-12700H (6 P-cores, 8 E-cores, total 20 threads), 64GB RAM each, 1TB NVMe, two 2.5Gbps network interfaces per node. One network interface is dedicated to the storage network (VLAN200), the other network interface hosts the mgmt-network, an untagged network and 2 VLANs. All interfaces are connected to 2.5Gbps switches.

I hope this post might help others having the same problem.

**Comment by jrlittlefoot64** [comment]
I had the same behavior and after reboot of this node the installation finished without errors

**Comment by Yu-Jack** [comment]
@bk201 I think this one is related https://github.com/harvester/harvester/issues/6041 because the OP of #6041 end up getting two succeeded and one pre-drained node, which is similar case to this issue.

**Comment by bk201** [comment]
Hi Jack, it's a bit different. 6041's symptom always happens on the first node (from the reports we get).

**Comment by Yu-Jack** [comment]
@bk201 Forget to add comment link, I mean this comment https://github.com/harvester/harvester/issues/6041#issuecomment-2180319204

**Comment by W0n9** [comment]
It is very similar to my [problem.](https://github.com/harvester/harvester/issues/6041#issuecomment-2180319204) 
But my symptom is the last node stack at pre-drainED, and it didn't turn to post-drain.
From `kubectl get vmi -n harvester-system`, I got only 1 vmi is running, and I tried to delete the pod related to this vmi, it doesn't work. Even reboot this node, it still stuck at pre-drainED.

**Comment by Yu-Jack** [comment]
[Update-1]

There is an error message about instance-manager in cattle-system/rancher-xxxx pod. **It provides some useful messages when upgrading.**

```
2024-07-02T02:50:01.556273130Z error when evicting pods/"instance-manager-7389ba963cbdeaa51e8748f5425ef7c2" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
2024-07-02T02:50:01.556301315Z error when evicting pods/"instance-manager-eef886d793d4e3723bae6a353ca55c43" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

Then I try to delete some unused PVC, then it successfully evict the instance-manager. Now, the stuck node starts to reboot.

---

[Original Post]
I just encountered similar issue when I'm doing experiment, but my first node is stuck in pre-drained mode. My case is my some pods don't match pod affinity, message in resource status is `0/2 nodes are available: 1 node(s) didn''t match pod anti-affinity rules, 1 node(s) were unschedulable. preemption: 0/2 nodes are available: 1 No preemption victims found for incoming pod, 1 Preemption is not helpful for scheduling..`. My one of nodes is tainted to `kubevirt.io/drain=draining:NoSchedule`, so there is no any pod could be scheduled to this node. The pod of vmi which OP mentioned seems run well.

Environment is two nodes installed in kvm (v1.3.0). Attachments are support bundle and upgrade log.

![image](https://github.com/harvester/harvester/assets/6960289/14da90bc-73cb-468f-be8f-fdad7f11cc82)

[supportbundle_42926605-82b1-4260-bdf7-fc51010466cf_2024-07-02T02-49-45Z.zip](https://github.com/user-attachments/files/16061051/supportbundle_42926605-82b1-4260-bdf7-fc51010466cf_2024-07-02T02-49-45Z.zip)
[hvst-upgrade-w5skg-upgradelog-archive-2024-07-02T02-53-30Z.zip](https://github.com/user-attachments/files/16061052/hvst-upgrade-w5skg-upgradelog-archive-2024-07-02T02-53-30Z.zip)

**Comment by irishgordo** [comment]
I have encountered this on a two node cluster, bare-metal, HP DL160 (v1.3.0->v1.3.1).

The manual workaround that needed to be implemented was:

[misconfigured longhorn instance manager pod disruption budget](https://docs.harvesterhci.io/v1.3/upgrade/v1-1-2-to-v1-2-0/#52-misconfigured-longhorn-instance-manager-r-pod-disruption-budgets-pdb)
Though, to note, based on that guide, I disregarded any volumes that were "degraded", which deviates from the guide's outlined steps slightly.

Once that was implemented, the first node shifted from PreDrained -> PostDraining -> Waiting Reboot, etc.

The workaround, did not need to be re-implemented on the second node.
